### PR TITLE
ILE-2 Delivery Employee Backend

### PR DESCRIPTION
**Pull request by dylanmckk**

### CODE REVIEW

**Review by CharlotteMKainos — APPROVED** [review]

File: src/main/java/org/example/TestApplication.java [hunk 1]
```diff
@@ -6,8 +6,12 @@
 import io.federecio.dropwizard.swagger.SwaggerBundle;
```
Comment (lukemKainos) [inline src/main/java/org/example/TestApplication.java:6]: its fine, i just dont see any logging, might be something to consider

File: src/main/java/org/example/controllers/DeliveryEmployeeController.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        try {
+            return Response.ok().entity(
+                    deliveryEmployeeService.getAllDeliveryEmployees()
+            ).build();
+        } catch (SQLException e) {
+            return Response.status(
+                    Response.Status.INTERNAL_SERVER_ERROR
+            ).build();
+        }
+    }
```
Comment (lukemKainos) [inline src/main/java/org/example/controllers/DeliveryEmployeeController.java:44]: space

File: src/main/java/org/example/controllers/DeliveryEmployeeController.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                    deliveryEmployeeService.getDeliveryEmployeeById(id)
+            ).build();
+        } catch (SQLException e) {
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+                    .entity(e.getMessage()).build();
+        } catch (DoesNotExistException e) {
+            return Response.status(Response.Status.NOT_FOUND)
+                    .entity(e.getMessage()).build();
+        }
+
```
Comment (lukemKainos) [inline src/main/java/org/example/controllers/DeliveryEmployeeController.java:61]: space?

File: src/main/java/org/example/daos/DeliveryEmployeeDao.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        statement.setString(3, deliveryEmployeeRequest.getBankAccountNumber());
+        statement.setString(4,
+                deliveryEmployeeRequest.getNationalInsuranceNumber());
+        statement.setInt(5, id);
+
+        statement.executeUpdate();
+    }
+
+    public void deleteDeliveryEmployee(final int id) throws SQLException {
+        Connection c = DatabaseConnector.getConnection();
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/DeliveryEmployeeDao.java]: this isnt closed

File: src/main/java/org/example/daos/DeliveryEmployeeDao.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            return resultSet.getInt(1);
+        }
+
+        return -1;
+    }
+
+    public void updateDeliveryEmployee(final int id,
+                                       final DeliveryEmployeeRequest deliveryEmployeeRequest)
+            throws SQLException {
+        Connection c = DatabaseConnector.getConnection();
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/DeliveryEmployeeDao.java]: this isnt closed

File: src/main/java/org/example/daos/DeliveryEmployeeDao.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                        resultSet.getString("nationalInsuranceNumber")
+                );
+            }
+        }
+        return null;
+    }
+
+    public int createDeliveryEmployee(DeliveryEmployeeRequest deliveryEmployee)
+            throws SQLException {
+        Connection c = DatabaseConnector.getConnection();
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/DeliveryEmployeeDao.java]: this isnt closed

File: src/main/java/org/example/daos/DeliveryEmployeeDao.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+                deliveryEmployees.add(deliveryEmployee);
+            }
+        }
+        return deliveryEmployees;
+    }
+
+    public DeliveryEmployee getDeliveryEmployeeById(int id)
+            throws SQLException {
+        try (Connection connection = DatabaseConnector.getConnection()) {
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/DeliveryEmployeeDao.java:41]: i like this connection inside a try, how about using https://stackoverflow.com/questions/26516020/try-with-resources-vs-try-catch for the rest?

File: src/main/java/org/example/exceptions/DoesNotExistException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package org.example.exceptions;
+
+public class DoesNotExistException extends Exception {
+    public DoesNotExistException(final Entity entity) {
+        super(entity.getEntity() + " does not exist");
```
Comment (lukemKainos) [inline src/main/java/org/example/exceptions/DoesNotExistException.java]: not a lot in this error message, perhaps something like "Delivery Employee" + entity + "does not exist?

File: src/main/java/org/example/exceptions/Entity.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package org.example.exceptions;
+
+public enum Entity {
+    DELIVERYEMPLOYEE("Delivery Employee");
```
Comment (lukemKainos) [inline src/main/java/org/example/exceptions/Entity.java]: enums are written by DELIVERY_EMPLOYEE

File: src/main/java/org/example/mappers/DeliveryEmployeeMapper.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
+package org.example.mappers;
+
+import org.example.models.DeliveryEmployee;
+import org.example.responses.DeliveryEmployeeResponse;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DeliveryEmployeeMapper {public static DeliveryEmployeeResponse mapDeliveryEmployeeToDeliveryEmployeeResponse(
```
Comment (lukemKainos) [inline src/main/java/org/example/mappers/DeliveryEmployeeMapper.java]: should be a new line here

File: src/main/java/org/example/models/DeliveryEmployee.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    public void setDeliveryEmployeeId(int deliveryEmployeeId) {
+        this.deliveryEmployeeId = deliveryEmployeeId;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
```
Comment (lukemKainos) [inline src/main/java/org/example/models/DeliveryEmployee.java]: could this be final?

File: src/main/java/org/example/models/DeliveryEmployeeRequest.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        this.salary = salary;
+        this.bankAccountNumber = bankAccountNumber;
+        this.nationalInsuranceNumber = nationalInsuranceNumber;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
```
Comment (lukemKainos) [inline src/main/java/org/example/models/DeliveryEmployeeRequest.java]: again final?

File: src/main/java/org/example/services/DeliveryEmployeeService.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.sql.SQLException;
+import java.util.List;
+
+public class DeliveryEmployeeService {
+    DeliveryEmployeeDao deliveryEmployeeDao;
+    DeliveryEmployeeValidator deliveryEmployeeValidator;
+
+    public DeliveryEmployeeService(final DeliveryEmployeeDao deliveryEmployeeDao, final DeliveryEmployeeValidator deliveryEmployeeValidator) {
+    this.deliveryEmployeeDao = new DeliveryEmployeeDao();
```
Comment (lukemKainos) [inline src/main/java/org/example/services/DeliveryEmployeeService.java]: instantiating a class being passed through?

File: src/main/java/org/example/services/DeliveryEmployeeService.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    public List<DeliveryEmployeeResponse> getAllDeliveryEmployees() throws SQLException {
+        return DeliveryEmployeeMapper.mapDeliveryEmployeeListToDeliveryEmployeeResponseList(
+                deliveryEmployeeDao.getAllDeliveryEmployees());
+    }
+
+    public DeliveryEmployeeResponse getDeliveryEmployeeById(final int id)
+            throws SQLException, DoesNotExistException {
+        DeliveryEmployee deliveryEmployee = deliveryEmployeeDao.getDeliveryEmployeeById(id);
+
+        if (deliveryEmployee == null) {
```
Comment (lukemKainos) [inline src/main/java/org/example/services/DeliveryEmployeeService.java:39]: does this pass a null or empty check?

File: src/main/java/org/example/services/DeliveryEmployeeService.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        return DeliveryEmployeeMapper.mapDeliveryEmployeeToDeliveryEmployeeResponse(deliveryEmployee);
+    }
+
+    public int createDeliveryEmployee(final DeliveryEmployeeRequest deliveryEmployeeRequest)
+        throws FailedToCreateException, SQLException, InvalidException {
+        deliveryEmployeeValidator.validateDeliveryEmployee(deliveryEmployeeRequest);
+
+        int id = deliveryEmployeeDao.createDeliveryEmployee(deliveryEmployeeRequest);
```
Comment (lukemKainos) [inline src/main/java/org/example/services/DeliveryEmployeeService.java]: final

File: src/main/java/org/example/validators/DeliveryEmployeeValidator.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        if (deliveryEmployeeRequest.getSalary() < 18000 ) {
+            throw new InvalidException(Entity.DELIVERYEMPLOYEE, "Salary must be above 18000 to meet the minimum wage.");
+        }
+
+        if (deliveryEmployeeRequest.getBankAccountNumber().length() > 34) {
+            throw new InvalidException(Entity.DELIVERYEMPLOYEE, "Bank account number greater than 34 characters.");
+        }
+
+        if (deliveryEmployeeRequest.getNationalInsuranceNumber().length() < 9) {
+            throw new InvalidException(Entity.DELIVERYEMPLOYEE, "National Insurance number greater than 9 characters.");
```
Comment (lukemKainos) [inline src/main/java/org/example/validators/DeliveryEmployeeValidator.java]: thats not the right message

File: src/main/java/org/example/validators/DeliveryEmployeeValidator.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package org.example.validators;
+
+import org.example.exceptions.Entity;
+import org.example.exceptions.InvalidException;
+import org.example.models.DeliveryEmployeeRequest;
+
+public class DeliveryEmployeeValidator {
+    public void validateDeliveryEmployee(
```
Comment (lukemKainos) [inline src/main/java/org/example/validators/DeliveryEmployeeValidator.java:8]: what happens if any field is null?

**Review by lukemKainos — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/services/DeliveryEmployeeService.java (hunk 14 above)
Comment (lukemKainos) [reply src/main/java/org/example/services/DeliveryEmployeeService.java:39]: it doesnt run validate here so it still doesnt pass empty check

File: src/main/java/org/example/validators/DeliveryEmployeeValidator.java (hunk 17 above)
Comment (lukemKainos) [reply src/main/java/org/example/validators/DeliveryEmployeeValidator.java:8]: not sure how you resplved

**Review by lukemKainos — APPROVED** [review]

**Review by dana-kainos — APPROVED** [review]
lots of love

**Review by CharlotteMKainos — APPROVED** [review]

**Pull request merged**